rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,230 @@
+
+When the DNS Operator publishes any changes to a zone, it will also add/update a (`kuadrant.last_modified_at`) TXT record with the current timestamp.
+
+If the DNS Operator has no local zoneRecords for a zone, it will list and build the zoneRecords.
+
+If the DNS Operator already has zoneRecords (and their lastModified time is older than the refreshInterval, more on this later) it will first execute a GET request for the `kuadrant.last_modified_at` TXT record, if the value of that field is more recent than the zoneRecords lastModified time, then the DNS Operator will list and rebuild it's zoneRecords.
+
+If the zoneRecords lastModified time is less old than the refreshInterval, a GET and possibly a LIST should only be executed when the spec of the kuadrantRecord has changed (as this will likely result in a write to the API).
+
+#### Ensure local records are accurate
+
+Once the zoneRecord is created / updated on cluster, the operator will then scan through this zoneRecord and remove any records for the local cluster. Then inject the values from the kuadrantRecord into the zoneRecord. After this the DNS Operator will sanity check the result and prune any bad records (we will expand on that later). Once the zoneRecord is updated and sanity checked, if it has changed then the result is sent back to the DNS Provider so that the zone can be updated and made accurate (with an added update to the `kuadrant.last_modified_at`) TXT record.
+
+#### Update response to failing health check 
+
+After ensuring its own ClusterID is present in the relevant TXT record, the controller will count the totalClusters which is the sum of unique clusterIDs in that hosts DNS records. Then it will count the reportingClusters which is the number of clusterIDs in the relevant TXT record, if `reportingClusters/totalClusters >= HealthCheck.MultiClusterFailureThreshold` then this IP will be removed for this host.
+
+example healthcheck TXT record:
+```
+myapp.pb.apps.com/24.34.45.56: ghye7d34; lhoyg391
+```
+
+As this TXT record can be quite long (when many clusters are present), if this fields exceeds a DNS Record limit (255 characters), it will run into a second TXT record with the same key.
+
+Example multiple healthcheck TXT record:
+```
+myapp.pb.apps.com/24.34.45.56: ghye7d34; lhoyg391; ... <more clusterIDs>
+myapp.pb.apps.com/24.34.45.56: hir01fcd; aoe8912;
+```
+
+##### Handling no IPs published
+
+If after processing the health checks the result is a DNS Record with no A records, then all of the A records will be persisted (this is to avoid an NXDOMAIN response).
+
+#### Regularly reconcile DNS Records
+
+As changes to the zone file can happen without triggering an event in the DNS Operator, the operator will need to poll the DNS Provider in order to ensure that it has an accurate set of records to ensure that changes in the zone are not missed. e.g. a cluster would not have health checks for any clusters that were added since it last read the zone.
```

Comment:
```suggestion
As changes to the zone file can happen without triggering an event in the DNS Operator, the operator will need to poll the DNS Provider in order to ensure that it has an accurate set of records. e.g. a cluster would not have health checks for any clusters that were added since it last read the zone.
```